File: src/main/java/visitor/Counter.java
package visitor;

import calculator.Expression;
import calculator.MyNumber;
import calculator.Operation;

public class Counter extends Visitor {

    private int depth = 0;
    private int ops = 0;
    private int nbs = 0;

    @Override
    public void visit(MyNumber n) {
        nbs++;
    }

    @Override
    public void visit(Operation o) {
        int maxDepth = o.getArgs().stream()
                .mapToInt(Expression::countDepth)
                .max()
                .orElse(0);
        depth = Math.max(depth, maxDepth + 1);
        ops++;
        o.setDepth(depth);
        o.setOps(ops);
        o.setNbs(nbs);
    }
}

File: src/main/python/visitor/counter.py
from streamerate import stream

from src.main.python.visitor.visitor import Visitor


class Counter(Visitor):

    def __init__(self):
        self.__depth = 0
        self.__ops = 0
        self.__nbs = 0

    def visit_my_number(self, number):
        self.__nbs += 1

    def visit_operation(self, o):
        max_depth = stream(o.get_args).map(lambda e: e.get_depth()).max()
        self.__depth = max(self.__depth, max_depth + 1)
        self.__ops += 1
        o.set_depth(self.__depth)
        o.set_ops(self.__ops)
        o.set_nbs(self.__nbs)

    def visit_linear_solution(self, solution):
        pass

    def visit_matrix(self, matrix):
        """The Visitor can traverse a matrix"""
        pass
